### ISSUE TITLE
UWP (Universal Windows Platform) support

### ISSUE THREAD

**Issue by FilippoCinotti** [keijiro/jp.keijiro.apriltag#3]
I'd like to pass to UWP to run this project on windows HoloLens 2, but I can't figure out how to solve these errors:
error CS0246: The type or namespace name 'Span<>' could not be found (are you missing a using directive or an assembly reference?)

Can you help me?

**Comment by keijiro** [comment]
I don't have any experience/knowledge on UWP/HoloLens, so I have no idea about the issue.

I keep this issue open for future reference, but please don't expect  me to fix it in the near future.

**Comment by FilippoCinotti** [comment]
Thank you so much! A second question... yuo estimate apriltag pose in local coordinates (in respect to camera position), how should I pass to worl coord? Thanks in advance

**Comment by keijiro** [comment]
That depends on how you define the world coordinate. I think it's a kind of application-specific problem.

**Comment by keijiro** [comment]
By the way, please don't reuse the thread for other issues. I'll only discuss the topic (UWP support) in this issue page.

**Comment by ghost** [comment]
@FilippoCinotti I also tried to use the project for the Hololens 2 and I bypassed this issue by importing the right packages manually (System.Buffers, System.Memory, ... initially imported by nuget if I'm right) to the project.
However I did not success to compile the dll with aarch64-linux-gnu with make for the Hololens2 requirement (UWP and ARM64) and I am stuck here...

**Comment by FilippoCinotti** [comment]
@Yafilo the same happen to me, let me know if you are able to make progresses

**Comment by ghost** [comment]
Hey so I found a repo to get the ARM64 of aprilTag through VCPKG : 
https://github.com/ooeygui/vcpkg (fork of the official vcpkg that has an ariltag implementation). Just install vcpkg as normal then install apriltag with : 
```.\vcpkg\vcpkg install apriltag```
It will search another fork of the official APRILTag repo where the threads are managed for windows and build an ARM 64 dll for UWP of april tag here : 
```.\vcpkg\buildtrees\apriltag\arm64-uwp-rel ```

Hope it helps

**Comment by FilippoCinotti** [comment]
Thanks @Yafilo, let's get in touch!

**Comment by FelixWeichselgartner** [comment]
@Yafilo Where do I have to execute the vcpkg? On my local computer in terminal or in Unity? How do I use it in Unity then. Sry but I'm a little confused. TY in advance!

**Comment by ghost** [comment]
You can download the repo and then launch ```bootstrap-vcpkg.bat``` It should install averything you need for vcpkg, Once it is done you can install apriltag package with ```.\vcpkg\vcpkg install apriltag```.

During the installation I got an hash error with a file, you can just modify the expected hash signature with the current haash signature. I don't remember wich file it was but it worked for me.

**Comment by FilippoCinotti** [comment]
Do you manage to complete it? I got a building error

![image](https://user-images.githubusercontent.com/63125253/118982706-f952c800-b97b-11eb-8c09-2b0f532fa0aa.png)

**Comment by FilippoCinotti** [comment]
@Yafilo @FelixWeichselgartner Do you manage to complete it? I got a building error

**Comment by FelixWeichselgartner** [comment]
I did not try so far, because I expected that would happen ^^.

**Comment by muxavir** [comment]
![image](https://user-images.githubusercontent.com/8582320/120430031-d10d9500-c38f-11eb-987d-8440600325ac.png)
Facing same error as above mentioned by @FilippoCinotti 
If anyone got chance to bypass this error, please share.

**Comment by FilippoCinotti** [comment]
@Yafilo I manage to solve the issue by substituting the correct hash code in the folder ./vcpkg/ports/apriltag/portfile (@FelixWeichselgartner @muxavir ). However, I only get x86 version of the library... how do you get this .\vcpkg\buildtrees\apriltag\arm64-uwp-rel ?

![image](https://user-images.githubusercontent.com/63125253/122918486-78626400-d35f-11eb-8be3-70e404d83288.png)

**Comment by ghost** [comment]
You can add parameters on building, if I remember well, you can set something like : 
`vcpkg install apriltag:arm64-uwp` for ARM64 or 
`vcpkg install apriltag:arm-uwp` for ARM

Hope it could help

**Comment by FilippoCinotti** [comment]
@Yafilo could you please upload the library?

yes the command should be `vcpkg install apriltag --triplet=arm64-uwp`

**Comment by FilippoCinotti** [comment]
But it gaves me an error

**Comment by muxavir** [comment]
any luck for arm64 @FilippoCinotti ?

**Comment by FilippoCinotti** [comment]
None :(

Il mar 6 lug 2021, 09:56 Musawar Arshad ***@***.***> ha
scritto:

**Comment by artificial-jamesirei** [comment]
@FilippoCinotti I am right there with you man, I'm trying to do the same thing.

I am able to build out a dll for arm64 by doing `vcpkg install apriltag --triplet=arm64-windows` but I'm not 100% certain that it's going to work when I build out for Hololens.  Can someone give me a quick ELI5 on how generating my own dll's with vcpkg works?  I'm kind of new to the whole process, and if I understand more I can try to generate one using the original repo perhaps

**Comment by markgrossnickle** [comment]
I was able to build the vcpkg package using the above methods. 

I grabbed the latest vcpkg and merged it with the branch from ooeygui. Updated the hash. Used the triplet code listed above for arm (haven't tried arm64 yet). 

But when I run it on the device I get a Access violation reading location....   It is in the assembly-csharp.cpp and muddled with IL2CPP code so its hard to really tell where its occurring other than  DetectionArray_get_Item_m is in the method name.  

Appears to be this:

```
    public unsafe ref Detection this[int i]
      => ref Util.AsRef<Detection>((void*)AsPointerArray.AsSpan[i]);
```

**Comment by markgrossnickle** [comment]
Figured out my issue. It at least no longer crashes and it does say it has found 1 detected tag (though I haven't tried drawing it in the spatial scene yet).  

In ZArray.cs the el_sz was a ulong and the branch I was working off of has el_sz in the zarray struct in apriltags.c as a size_t or uint. Thus the parsing/conversion was off.

**Comment by rubit0** [comment]
> Figured out my issue. It at least no longer crashes and it does say it has found 1 detected tag (though I haven't tried drawing it in the spatial scene yet).

Hi @markgrossnickle, would you be willing to share your .dll file so that we others don't have to go through entire build ordeal? I would appreciate it a lot!

**Comment by markgrossnickle** [comment]
Here is a zip of the Arm build. It also has a debug method which outputs text back to Unity
[apriltags.zip](https://github.com/keijiro/jp.keijiro.apriltag/files/9133880/apriltags.zip)

**Comment by theloni-monk** [comment]
Has anyone here tried to run this on the HoloLens 1? Its an Intel x86 so I hoped it would run the default DLLs but I am getting the following error:
 ```
DllNotFoundException: Unable to load DLL 'AprilTag'. Tried the load the following dynamic libraries: Unable to load dynamic library 'AprilTag' because of 'C:\buildslave\unity\build\External\il2cpp\builds\external\baselib\Source\WinApi\Baselib_DynamicLibrary_WinApi.inl.h(36):Baselib_DynamicLibrary_OpenUtf16: Failed to open the requested dynamic library (0x06000000) - The specified module could not be found.  (WinError:0000007e)
  at AprilTag.Interop.Detector._Create () [0x00000] in <00000000000000000000000000000000>:0 
  at AprilTag.Interop.Detector.Create () [0x00000] in <00000000000000000000000000000000>:0 
  at AprilTag.TagDetector..ctor (System.Int32 width, System.Int32 height, System.Int32 decimation) [0x00000] in <00000000000000000000000000000000>:0 
  at DetectionTest.Start () [0x00000] in <00000000000000000000000000000000>:0 
```
To be clear, this is a UWP support error, as it should work out of box with something like the HoloLens 1 which is an x86 UWP platform.

**Comment by N0RAIR** [comment]
> Here is a zip of the Arm build. It also has a debug method which outputs text back to Unity [apriltags.zip](https://github.com/keijiro/jp.keijiro.apriltag/files/9133880/apriltags.zip)

Hey @markgrossnickle , is there a possibility you can build and share the arm64 build too? the arm does not work for arm64 builds and i can't build for arm i need arm64 for the Hololens 2

@theloni-monk I get the same issue when I run it on the HoloLens 2
